File: src/FileContainer.js
var T    = require ( "./Tango" ) ;
var Path = require ( 'path') ;
var fs   = require ( 'fs') ;

/**
 * Description
 * @constructor
 * @return 
 */
var FileContainer = function ( file )
{
  this.className = "FileContainer" ;
  this.path = "" ;
  this.name = "" ;
  this.data = null ;
  if ( file )
  {
    this.path = file.replace ( /\\/g, "/" ) ;
    this.path = Path.resolve ( this.path ) ;
    this.path = this.path.replace ( /\\/g, "/" ) ;
    this.name = this.path.substring ( this.path.lastIndexOf ( "/" ) + 1 ) ;
  }
  this.targetIsLocalHost = false ;
};
/**
 * Description
 * @method toString
 * @return string
 */
FileContainer.prototype.toString = function()
{
  return "(" + this.className + ")[  \npath=" + this.path + "\n  name=" + this.name + "\n  data=" + this.data + "\n]" ;
};
FileContainer.prototype.setTargetIsLocalHost = function ( state )
{
  this.targetIsLocalHost = !!state ;
};
/**
 * Description
 * @return string 
 */
FileContainer.prototype.toJSON = function()
{
  var data = this.data ;
  if ( ! this.targetIsLocalHost && ! data )
  {
    data = fs.readFileSync ( this.path ) ;
  }
  delete this.targetIsLocalHost ;
  return { className:'FileContainer', path: this.path, name:this.name, data:data } ;
};
FileContainer.prototype.getBytes = function()
{
  if ( this.data )
  {
    return this.data ;
  }
  return fs.readFileSync ( this.path ) ;
};
FileContainer.prototype.getName = function()
{
  return this.name ;
};
FileContainer.prototype.getPath = function()
{
  return this.path ;
};
FileContainer.prototype.write = function ( fullFileName )
{
  var ws ;
  if ( ! this.data )
  {
    var options =
    {
      flags: 'r'
    , encoding: null
    , fd: null
    , autoClose: true
    } ;
    var stat = fs.statSync ( this.path ) ;
    ws = fs.createWriteStream ( fullFileName, { encoding: null } ) ;
    var rs = fs.createReadStream ( this.path, options ) ;
    rs.pipe ( ws ) ;
  }
  else
  {
    ws = fs.createWriteStream ( fullFileName, { encoding: null } ) ;
    ws.write ( this.data ) ;
    ws.end() ;
  }
};

module.exports = FileContainer ;

File: java/src/org/gessinger/gepard/FileContainer.java
package org.gessinger.gepard ;

import java.util.* ;
import java.io.* ;
import java.text.* ;

public class FileContainer implements JSONEncodable, JSONDecodable, HasSetTargetIsLocalHost
{
	String path                         = "" ;
	String name                         = "" ;
	byte[] data                         = null ;
	transient boolean targetIsLocalHost = false ;
	public FileContainer ( String file )
	throws Exception
	{
	  if ( file != null )
	  {
	    path = file.replace ( '\\', '/' ) ;
	    path = new File ( path ).getCanonicalPath() ;
	    path = path.replace ( '\\', '/' ) ;
	    name = path.substring ( path.lastIndexOf ( "/" ) + 1 ) ;
	  }
	}
	public FileContainer()
	throws Exception
	{
	}
	public String toString()
	{
	  return "(" + this.getClass().getName() + ")[\n  path=" + path + "\n  name=" + name + "\n  date=" + Util.toString ( data ) + "\n]" ;
	}
	public void setTargetIsLocalHost ( boolean state )
	{
	  targetIsLocalHost = state ;
	}
	public void fromJSON ( Map<String,Object> map )
	throws IOException
	{
		path = (String) map.get ( "path" ) ;
		name = (String) map.get ( "name" ) ;
		data = (byte[]) map.get ( "data" ) ;
	}
	public Map<String,Object> toJSON()
	throws IOException
	{
		byte[] d = this.data ;
		if ( ! targetIsLocalHost && data == null )
		{
		  data = getBytes() ;
		}
		HashMap<String,Object> hm = new HashMap<String,Object>() ;
		hm.put ( "className", "FileContainer" ) ;
		hm.put ( "path", path ) ;
		hm.put ( "name", name ) ;
		hm.put ( "data", data ) ;
    Util.convertJavaTypedDataToNodeJS ( targetIsLocalHost, hm ) ;
		return hm ;
	}
	public byte[] getBytes()
	throws IOException
	{
	  if ( data != null )
	  {
	    return data ;
	  }
	  return Util.getBytes ( new File ( path ) ) ;
	}
	public String getName()
	{
	  return this.name ;
	};
	public String getPath()
	{
	  return this.path ;
	};
	public void write ( String fullFileName )
	throws Exception
	{
		InputStream in = null ;
		OutputStream out = null ;
		if ( data != null )
		{
			in = new ByteArrayInputStream ( data ) ;
		}
		else
		{
			in = new FileInputStream ( path ) ;
		}
		try
		{
			out = new FileOutputStream ( fullFileName ) ;
			Util.copy ( in, out ) ;
		}
		catch ( Exception exc )
		{
			System.err.println ( Util.toString ( exc ) ) ;
			throw exc ;
		}
    finally
    {
      if ( in != null ) try { in.close() ; } catch ( Exception e ) {}
      if ( out != null ) try { out.close() ; } catch ( Exception e ) {}
    }
	}
}
